diff -Nur mount.orig/mount.c mount/mount.c
--- mount.orig/mount.c	2021-12-11 09:26:50.072144000 -0700
+++ mount/mount.c	2022-01-14 10:09:55.243655000 -0700
@@ -261,15 +261,15 @@
 	struct fstab *fs;
 	struct statfs *mntbuf;
 	int all, ch, i, init_flags, late, failok, mntsize, rval, have_fstab, ro;
-	int onlylate;
+	int onlylate, nohupmountd;
 	char *cp, *ep, *options;
 
-	all = init_flags = late = onlylate = 0;
+	all = init_flags = late = onlylate = nohupmountd = 0;
 	ro = 0;
 	options = NULL;
 	vfslist = NULL;
 	vfstype = "ufs";
-	while ((ch = getopt(argc, argv, "adF:fLlno:prt:uvw")) != -1)
+	while ((ch = getopt(argc, argv, "adF:fLlNno:prt:uvw")) != -1)
 		switch (ch) {
 		case 'a':
 			all = 1;
@@ -290,6 +290,9 @@
 		case 'l':
 			late = 1;
 			break;
+		case 'N':
+			nohupmountd = 1;
+			break;
 		case 'n':
 			/* For compatibility with the Linux version of mount. */
 			break;
@@ -482,7 +485,7 @@
 	 * If the mount was successfully, and done by root, tell mountd the
 	 * good news.
 	 */
-	if (rval == 0 && getuid() == 0)
+	if (rval == 0 && getuid() == 0 && !nohupmountd)
 		restart_mountd();
 
 	exit(rval);
